scene/resources/particle_process_material.cpp

```diff
@@ -988,13 +988,11 @@ void ParticleProcessMaterial::_update_shader() {
 	code += "	\n";
 	if (collision_mode == COLLISION_RIGID) {
 		code += "	if (COLLIDED) {\n";
-		code += "		if (length(VELOCITY) > 3.0) {\n";
-		code += "			TRANSFORM[3].xyz += COLLISION_NORMAL * COLLISION_DEPTH;\n";
-		code += "			VELOCITY -= COLLISION_NORMAL * dot(COLLISION_NORMAL, VELOCITY) * (1.0 + collision_bounce);\n";
-		code += "			VELOCITY = mix(VELOCITY,vec3(0.0),clamp(collision_friction, 0.0, 1.0));\n";
-		code += "		} else {\n";
-		code += "			VELOCITY = vec3(0.0);\n";
-		code += "		}\n";
+		code += "		float collision_response = dot(COLLISION_NORMAL, VELOCITY);\n";
+		code += "		float collision_trigger = step(2, length(collision_response)/clamp(collision_bounce + 1.0, 1.0, 2.0));\n";
+		code += "		TRANSFORM[3].xyz += COLLISION_NORMAL * COLLISION_DEPTH;\n";
+		code += "		VELOCITY -= COLLISION_NORMAL * collision_response * (1.0 + collision_bounce * collision_trigger);\n";
+		code += "		VELOCITY = mix(VELOCITY,vec3(0.0),clamp(collision_friction, 0.0, 1.0));\n";
```

Comment:
I am not happy with the fact that friction is always applied. IMO the collision friction should depend on the direction the particle is trying to move, so that when the particle is trying to move e.g. straight away from the surface friction should be 0.

Reply:
I have done this by basically splitting the VELOCITY into 3 steps:
 1. Remove the component of the velocity pushing the particle further into the surface.
 2. Apply friction
 3. Add bounce component of velocity